src/filesystem.cpp

```diff
@@ -0,0 +1,256 @@
+#include "filesystem.h"
+#include "debug.h"
+
+#include <cstring>
+#include <string>
+#include <vector>
+#include <deque>
+#include <algorithm>
+#include <memory>
+
+// FILE I/O
+#include <sys/stat.h>
+#ifdef _MSC_VER
+#include "wdirent.h"
+#include <direct.h>
+#else
+#include <dirent.h>
+#include <unistd.h>
+#endif
+
+namespace {
+
+#if (defined _WIN32 || defined __WIN32__)
+bool do_mkdir(std::string const& path, int const mode)
+{
+    (void)mode; //not used on windows
+    return mkdir(path.c_str()) == 0;
+}
+#else
+bool do_mkdir(std::string const& path, int const mode)
+{
+    return mkdir(path.c_str(), mode) == 0;
+}
+#endif
+
+} //anonymous namespace
+
+bool assure_dir_exist(std::string const &path)
+{
+    std::unique_ptr<DIR, decltype(&closedir)> dir {opendir(path.c_str()), closedir};
+    return dir || do_mkdir(path, 0777);
+}
+
+bool file_exist(const std::string &path)
+{
+    struct stat buffer;
+    return ( stat( path.c_str(), &buffer ) == 0 );
+}
+
+#if (defined _WIN32 || defined __WIN32__)
+bool remove_file(const std::string &path)
+{
+    return DeleteFile(path.c_str()) != 0;
+}
+#else
+bool remove_file(const std::string &path)
+{
+    return unlink(path.c_str()) == 0;
+}
+#endif
+
+#if (defined _WIN32 || defined __WIN32__)
+bool rename_file(const std::string &old_path, const std::string &new_path)
+{
+    // Windows rename function does not override existing targets, so we
+    // have to remove the target to make it compatible with the linux rename
+    if (file_exist(new_path)) {
+        if(!remove_file(new_path)) {
+            return false;
+        }
+    }
+
+    return rename(old_path.c_str(), new_path.c_str()) == 0;
+}
+#else
+bool rename_file(const std::string &old_path, const std::string &new_path)
+{
+    return rename(old_path.c_str(), new_path.c_str()) == 0;
+}
+#endif
+
+namespace {
+
+//TODO move elsewhere.
+template <typename T, size_t N>
+inline size_t sizeof_array(T const (&)[N]) noexcept {
+    return N;
+}
+
+//--------------------------------------------------------------------------------------------------
+// For non-empty path, call function for each file at path.
+//--------------------------------------------------------------------------------------------------
+template <typename Function>
+void for_each_dir_entry(std::string const &path, Function function)
+{
+    using dir_ptr = std::unique_ptr<DIR, decltype(&closedir)>;
+
+    if (path.empty()) {
+        return;
+    }
+
+    dir_ptr root {opendir(path.c_str()), closedir};
+    if (!root) {
+        auto const e = errno;
+        DebugLog(DebugLevel::D_WARNING, DebugClass::D_MAIN) << "Warning: error " << e
+            << " couldn't open file " << path.c_str();
```

Comment:
Yep, missed that when I flipped the msg over to the stream version.
I'll also add strerror output as well.

Thanks for catching those.